pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1352,6 +1343,274 @@ func (vs *VSphere) NodeExists(c *govmomi.Client, nodeName k8stypes.NodeName) (bo
+	var newSCSIController types.BaseVirtualDevice
+	var newSCSICreated = false
+	if scsiController == nil {
+		newSCSIController, err = createandAttachSCSIControllerToVM(ctx, virtualMachine, diskControllerType)
+		if err != nil {
+			glog.Errorf("Failed to create SCSI controller for VM :%q with err: %+v", virtualMachine.Name(), err)
+			return err
+		}
+
+		// verify scsi controller in virtual machine
+		vmDevices, err := virtualMachine.Device(ctx)
+		if err != nil {
+			return err
+		}
+		scsiController = getSCSIController(vmDevices, diskControllerType)
+		if scsiController == nil {
+			glog.Errorf("cannot find SCSI controller in VM")
+			// attempt clean up of scsi controller
+			cleanUpController(ctx, newSCSIController, vmDevices, virtualMachine)
+			return fmt.Errorf("cannot find SCSI controller in VM")
+		}
+		newSCSICreated = true
+	}
+
+	kubeVolsPath := filepath.Clean(datastore.Path(VolDir)) + "/"
+	// Create a kubevols directory in the datastore if one doesn't exist.
+	err = makeDirectoryInDatastore(vs.client, datacenter, kubeVolsPath, false)
+	if err != nil && err != ErrFileAlreadyExist {
+		glog.Errorf("Cannot create dir %#v. err %s", kubeVolsPath, err)
+		return err
+	}
+
+	glog.V(4).Infof("Created dir with path as %+q", kubeVolsPath)
+
+	vmDiskPath := kubeVolsPath + volumeOptions.Name + ".vmdk"
+	disk := vmDevices.CreateDisk(scsiController, datastore.Reference(), vmDiskPath)
+	unitNumber, err := getNextUnitNumber(vmDevices, scsiController)
+	if err != nil {
+		glog.Errorf("cannot attach disk to VM, limit reached - %v.", err)
+		return err
+	}
+	*disk.UnitNumber = unitNumber
+	disk.CapacityInKB = int64(volumeOptions.CapacityKB)
+
+	backing := disk.Backing.(*types.VirtualDiskFlatVer2BackingInfo)
+	backing.DiskMode = string(types.VirtualDiskModeIndependent_persistent)
+
+	if diskFormat == ThinDiskType {
+		backing.ThinProvisioned = types.NewBool(true)
+	} else if diskFormat == EagerZeroedThickDiskType {
+		backing.EagerlyScrub = types.NewBool(true)
+	} else {
+		backing.ThinProvisioned = types.NewBool(false)
+	}
+
+	// Reconfigure VM
+	virtualMachineConfigSpec := types.VirtualMachineConfigSpec{}
+	deviceConfigSpec := &types.VirtualDeviceConfigSpec{
+		Device:        disk,
+		Operation:     types.VirtualDeviceConfigSpecOperationAdd,
+		FileOperation: types.VirtualDeviceConfigSpecFileOperationCreate,
+	}
+
+	storageProfileSpec := &types.VirtualMachineDefinedProfileSpec{
+		ProfileId: "",
+		ProfileData: &types.VirtualMachineProfileRawData{
+			ExtensionKey: "com.vmware.vim.sps",
+			ObjectData:   volumeOptions.StorageProfileData,
+		},
+	}
+
+	deviceConfigSpec.Profile = append(deviceConfigSpec.Profile, storageProfileSpec)
+	virtualMachineConfigSpec.DeviceChange = append(virtualMachineConfigSpec.DeviceChange, deviceConfigSpec)
+	task, err := virtualMachine.Reconfigure(ctx, virtualMachineConfigSpec)
+	if err != nil {
+		glog.Errorf("Failed to reconfigure the VM with the disk with err - %v.", err)
+		if newSCSICreated {
```

Comment:
I think this is not needed. If you plan to keep VM around then you can use same controller next time.